tokio/src/io/driver/scheduled_io.rs

```diff
@@ -1,28 +1,106 @@
-use crate::loom::future::AtomicWaker;
+use super::{platform, Direction, ReadyEvent, Tick};
 use crate::loom::sync::atomic::AtomicUsize;
+use crate::loom::sync::Mutex;
+use crate::util::bit;
 use crate::util::slab::Entry;
 
 use std::sync::atomic::Ordering::{AcqRel, Acquire, Release};
+use std::task::{Context, Poll, Waker};
+
+cfg_io_readiness! {
+    use crate::util::linked_list::{self, LinkedList};
+
+    use std::cell::UnsafeCell;
+    use std::future::Future;
+    use std::marker::PhantomPinned;
+    use std::pin::Pin;
+    use std::ptr::NonNull;
+}
 
 /// Stored in the I/O driver resource slab.
 #[derive(Debug)]
 pub(crate) struct ScheduledIo {
     /// Packs the resource's readiness with the resource's generation.
     readiness: AtomicUsize,
 
-    /// Task waiting on read readiness
-    pub(crate) reader: AtomicWaker,
+    waiters: Mutex<Waiters>,
+}
+
+#[cfg(feature = "io-readiness")]
+type WaitList = LinkedList<Waiter, <Waiter as linked_list::Link>::Target>;
+
+#[derive(Debug, Default)]
+struct Waiters {
+    #[cfg(feature = "io-readiness")]
+    /// List of all current waiters
+    list: WaitList,
+
+    /// Waker used for AsyncRead
+    reader: Option<Waker>,
+
+    /// Waker used for AsyncWrite
+    writer: Option<Waker>,
+}
+
+cfg_io_readiness! {
+    #[derive(Debug)]
+    struct Waiter {
+        pointers: linked_list::Pointers<Waiter>,
+
+        /// The waker for this task
+        waker: Option<Waker>,
```

Comment:
I _think_ you would have to put another `UnsafeCell` around this to make miri happy in the threaded scenario (assuming miri supports threads). This is because operations such as
```Rust
// Safety: called while locked
unsafe {
    (*waiter.get()).waker = Some(cx.waker().clone());
}
```
involve first creating a `&mut Waiter` and then creating a `&mut Option<Waker>` from that, but the creation of `&mut Waiter` asserts exclusive access to the entire struct.

Reply:
Could you elaborate on the problem? I believe this is fine, but if not, happy to fix.

Reply:
I created a [playground][1] with a similar situation. The first field is immutable and mirrors `interest`. The second field is mutated only from one place and mirrors `waiter`.

[1]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=efdd8a2d963c838d1e1cf1c827e27441

Reply:
The `interest` is copied, not referenced. When in `State::Init`, the waiter hasn't been put into the list and shared. And in `State::Waiting`, the lock is taken first.